File: ifitwala_ed/students/report/case_entries_activity_log/case_entries_activity_log.js
// Copyright (c) 2025, François de Ryckel and contributors
// For license information, please see license.txt

// ifitwala_ed/students/report/case_entries_activity_log/case_entries_activity_log.js

/* global frappe */

frappe.query_reports["Case Entries Activity Log"] = {
	// ---------- top filter bar ----------
	filters: [
		{ fieldname: "from_date",     label: __("From Date"),     fieldtype: "Date" },
		{ fieldname: "to_date",       label: __("To Date"),       fieldtype: "Date" },
		{ fieldname: "student",       label: __("Student"),       fieldtype: "Link",  options: "Student" },
		{ fieldname: "referral",      label: __("Referral"),      fieldtype: "Link",  options: "Student Referral" },
		{ fieldname: "school",        label: __("School"),        fieldtype: "Link",  options: "School" },
		{ fieldname: "program",       label: __("Program"),       fieldtype: "Link",  options: "Program" },
		{ fieldname: "academic_year", label: __("Academic Year"), fieldtype: "Link",  options: "Academic Year" },
		{ fieldname: "case_manager",  label: __("Case Manager"),  fieldtype: "Link",  options: "User" },
		{ fieldname: "assignee",      label: __("Entry Assignee"),fieldtype: "Link",  options: "User" },
		{ fieldname: "entry_type",    label: __("Entry Type"),    fieldtype: "Select",
			options: "\nMeeting\nCounseling Session\nAcademic Support\nCheck-in\nFamily Contact\nExternal Referral\nSafety Plan\nReview\nOther" },
		{ fieldname: "entry_status",  label: __("Entry Status"),  fieldtype: "Select",
			options: "\nOpen\nIn Progress\nDone\nCancelled" },
		{ fieldname: "case_severity", label: __("Case Severity"), fieldtype: "Select",
			options: "\nLow\nModerate\nHigh\nCritical" },
		{ fieldname: "case_status",   label: __("Case Status"),   fieldtype: "Select",
			options: "\nOpen\nIn Progress\nOn Hold\nEscalated\nClosed" },
		// Optional: initial chart bucket (the in-chart toggle can override this)
		{ fieldname: "time_bucket",   label: __("Time Bucket"),   fieldtype: "Select",
			options: "Week\nMonth", default: "Week" }
	],


	// ---------- header Print button ----------
	onload(report) {
		const page = report.page;
		ensure_print_button(page);
	},

	// ---------- extra charts with a Weekly/Monthly toggle ----------
	after_datatable_render(datatable) {
		// In v15 the hook receives the datatable instance; use the global query_report object.
		const page = (frappe.query_report && frappe.query_report.page) || null;
		if (!page || !page.main) return;

		ensure_print_button(page);

		const msg = (frappe.query_report && frappe.query_report.message) || {};

		const charts = {
			week: msg.chart_over_time_week || null,
			month: msg.chart_over_time_month || null,
			by_school: msg.chart_by_school || null,
			by_program: msg.chart_by_program || null,
			by_manager: msg.chart_by_manager || null
		};

		// Create/refresh wrapper (idempotent)
		let $area = page.main.find(".cea-extra-charts");
		if (!$area.length) {
			$area = $(`
				<div class="cea-extra-charts" style="margin-top: 1rem;">
					<div class="row gy-4">
						<!-- Time -->
						<div class="col-12 col-lg-6">
							<div class="d-flex align-items-center justify-content-between">
								<h5 class="mb-2">${__("Entries Over Time")}</h5>
								<div>
									<select class="form-select form-select-sm cea-bucket" style="width:auto;display:inline-block">
										<option value="week">${__("Weekly")}</option>
										<option value="month">${__("Monthly")}</option>
									</select>
								</div>
							</div>
							<div class="cea-chart cea-time"></div>
						</div>

						<!-- School / Program -->
						<div class="col-12 col-lg-6">
							<div class="d-flex align-items-center justify-content-between">
								<h5 class="mb-2">${__("Entries by School / Program")}</h5>
								<div>
									<select class="form-select form-select-sm cea-dim" style="width:auto;display:inline-block">
										<option value="school">${__("School")}</option>
										<option value="program">${__("Program")}</option>
									</select>
								</div>
							</div>
							<div class="cea-chart cea-dim-chart"></div>
						</div>

						<!-- By Case Manager -->
						<div class="col-12">
							<h5 class="mb-2">${__("Entries per Case Manager")}</h5>
							<div class="cea-chart cea-manager"></div>
						</div>
					</div>
				</div>
			`).appendTo(page.main);
		} else {
			// Clear chart containers on every re-render to avoid stacking canvases
			$area.find(".cea-chart").empty();
		}

		// Elements
		const $timeEl = $area.find(".cea-time")[0];
		const $dimEl = $area.find(".cea-dim-chart")[0];
		const $mgrEl = $area.find(".cea-manager")[0];
		const $bucket = $area.find(".cea-bucket");
		const $dimSel = $area.find(".cea-dim");

		function render_chart(el, spec) {
			if (!el) return;
			el.innerHTML = "";
			if (!spec || !spec.data || !Array.isArray(spec.data.labels) || !spec.data.labels.length) {
				el.innerHTML = `<div class="text-muted small">${__("No data")}</div>`;
				return;
			}
			new frappe.Chart(el, spec); // { data:{labels,datasets}, type:'line'|'bar' }
		}

		function current_time_spec(bucket) {
			const key = (bucket || "week").toLowerCase();
			return key === "month" ? (charts.month || charts.week) : (charts.week || charts.month);
		}

		function render_dim(which) {
			render_chart($dimEl, (which || "school") === "program" ? charts.by_program : charts.by_school);
		}

		// Initial selections (respect filter)
		const initBucket = (frappe.query_report.get_filter_value("time_bucket") || "Week").toLowerCase();
		$bucket.val(initBucket);
		render_chart($timeEl, current_time_spec(initBucket));
		render_dim($dimSel.val() || "school");
		render_chart($mgrEl, charts.by_manager);

		// Wire toggles (namespaced to avoid duplicates)
		$bucket.off("change.cea").on("change.cea", () => {
			render_chart($timeEl, current_time_spec($bucket.val()));
		});
		$dimSel.off("change.cea").on("change.cea", () => {
			render_dim($dimSel.val());
		});
	},
};


// ---------- helpers (print) ----------
function ensure_print_button(page) {
	const BTN_KEY = "cea-print-btn";

	// Avoid duplicates
	if (page.inner_toolbar && page.inner_toolbar.find(`button[data-key="${BTN_KEY}"]`).length) {
		return;
	}

	// Add a plain blue Print button (Bootstrap 4 → btn-info)
	const $btn = page.add_inner_button(__("Print"), () => handle_report_print(), null);
	if ($btn) {
		$btn.attr("data-key", BTN_KEY);
		$btn.removeClass("btn-default").addClass("btn-info");
	}
}

async function handle_report_print() {
	try {
		const report_name = "Case Entries Activity Log";
		const filters = (frappe.query_report && frappe.query_report.get_values)
			? frappe.query_report.get_values()
			: {};

		const r = await frappe.call({
			method: "frappe.desk.query_report.export_query",
			args: {
				report_name,
				file_format_type: "PDF",
				filters
			}
		});

		const url = r && r.message && r.message.file_url;
		if (url) {
			window.open(url, "_blank");
		} else {
			frappe.msgprint(__("Could not generate the PDF file for this report."));
		}
	} catch (e) {
		frappe.msgprint(__("Print failed. Please try the ⋮ menu or check permissions."));
	}
}

// ---------- helpers (client-side aggregates) ----------
function parse_rows(rows) {
	const out = {
		byWeek: {},
		byMonth: {},
		bySchool: {},
		byProgram: {},
		byManager: {}
	};

	rows.forEach(r => {
		const dateStr = (r.entry_date || r.entry_datetime || "").toString();
		if (dateStr) {
			const d = frappe.datetime.str_to_obj(dateStr);
			if (d) {
				const weekStart = new Date(d);
				weekStart.setDate(d.getDate() - ((d.getDay() + 6) % 7)); // Monday
				const wlabel = frappe.format(weekStart, "Date", { format: "d MMM yyyy" });
				out.byWeek[wlabel] = (out.byWeek[wlabel] || 0) + 1;

				const mStart = new Date(d.getFullYear(), d.getMonth(), 1);
				const mlabel = frappe.format(mStart, "Date", { format: "MMM yyyy" });
				out.byMonth[mlabel] = (out.byMonth[mlabel] || 0) + 1;
			}
		}

		const school = (r.school || "").trim() || __("—");
		out.bySchool[school] = (out.bySchool[school] || 0) + 1;

		const program = (r.program || "").trim() || __("—");
		out.byProgram[program] = (out.byProgram[program] || 0) + 1;

		const mgr = (r.case_manager_name || "").trim() || __("— Unassigned —");
		out.byManager[mgr] = (out.byManager[mgr] || 0) + 1;
	});

	return out;
}

function render_time(el, parsed, bucket) {
	const src = bucket === "month" ? parsed.byMonth : parsed.byWeek;
	const labels = Object.keys(src).sort((a, b) => new Date(a) - new Date(b));
	const values = labels.map(l => src[l]);

	el.innerHTML = "";
	new frappe.Chart(el, {
		data: { labels, datasets: [{ name: __("Entries"), values }] },
		type: "line",
		height: 240
	});
}

function render_dim(el, parsed, dim) {
	const src = dim === "program" ? parsed.byProgram : parsed.bySchool;
	const labels = Object.keys(src).sort();
	const values = labels.map(l => src[l]);

	el.innerHTML = "";
	new frappe.Chart(el, {
		data: { labels, datasets: [{ name: __("Entries"), values }] },
		type: "bar",
		height: 240
	});
}

function render_manager(el, parsed) {
	const entries = Object.entries(parsed.byManager).sort((a, b) => b[1] - a[1]).slice(0, 12);
	const labels = entries.map(x => x[0]);
	const values = entries.map(x => x[1]);

	el.innerHTML = "";
	new frappe.Chart(el, {
		data: { labels, datasets: [{ name: __("Entries"), values }] },
		type: "bar",
		height: 260
	});
}

File: ifitwala_ed/students/report/case_entries_activity_log/case_entries_activity_log.py
# Copyright (c) 2025, François de Ryckel and contributors
# For license information, please see license.txt

# ifitwala_ed/students/report/case_entries_activity_log/case_entries_activity_log.py

import re
from collections import defaultdict
from datetime import date, timedelta

import frappe
from frappe import _
from frappe.utils import formatdate, getdate, strip_html

ENTRY_LOCAL_EXPR = "CONVERT_TZ(rce.entry_datetime, 'UTC', %(site_tz)s)"


ALLOWED = {"Counselor", "Academic Admin"}


def execute(filters=None):
    filters = frappe._dict(filters or {})
    _guard_roles()
    site_tz = frappe.utils.get_system_timezone() or "UTC"

    # Dates (inclusive)
    date_from = getdate(filters.get("from_date")) if filters.get("from_date") else None
    date_to = getdate(filters.get("to_date")) if filters.get("to_date") else None

    where, params = _build_where(filters, date_from, date_to, site_tz)

    # ── Main query (note Employee join by user_id → employee_full_name) ──
    q = f"""
		SELECT
			{ENTRY_LOCAL_EXPR}                AS entry_datetime_local,
			rc.name                            AS referral_case,
			rc.referral                        AS referral,
			rc.student                         AS student,
			s.student_full_name                AS student_full_name,
			rc.school                          AS school,
			rc.program                         AS program,
			rc.academic_year                   AS academic_year,
			rc.case_manager                    AS case_manager,
			u.full_name                        AS user_full_name,
			e.employee_full_name               AS employee_full_name,
			rc.severity                        AS case_severity,
			rc.case_status                     AS case_status,
			rce.entry_type                     AS entry_type,
			rce.assignee                       AS assignee,
			rce.status                         AS entry_status,
			rce.summary                        AS summary_raw,

			/* NEW: latest activity per case within the filtered result set */
			MAX({ENTRY_LOCAL_EXPR}) OVER (PARTITION BY rc.name) AS _case_latest
		FROM `tabReferral Case` rc
		JOIN `tabReferral Case Entry` rce
					ON rce.parent = rc.name
		LEFT JOIN `tabStudent` s
					ON s.name = rc.student
		LEFT JOIN `tabUser` u
					ON u.name = rc.case_manager
		LEFT JOIN `tabEmployee` e
					ON e.user_id = rc.case_manager
		{("WHERE " + " AND ".join(where)) if where else ""}
		/* NEW: order cases by their latest activity, then entries newest-first */
		ORDER BY _case_latest DESC, rc.name ASC, entry_datetime_local DESC
	"""

    rows = frappe.db.sql(q, params, as_dict=True)

    # ── Presentational enrichments (no extra queries) ───────────────────
    for r in rows:
        # Pretty date (e.g., 9 September 2025)
        entry_dt = r.get("entry_datetime_local")
        entry_date = getdate(entry_dt) if entry_dt else None
        r["entry_date"] = entry_date
        r["entry_date_pretty"] = formatdate(entry_date, "d MMMM yyyy") if entry_date else ""

        # Linked student name
        full = r.get("student_full_name") or r.get("student") or ""
        sid = r.get("student") or ""
        r["student_link"] = (
            f'<a href="/app/student/{frappe.utils.escape_html(sid)}" target="_blank" rel="noopener">'
            f"{frappe.utils.escape_html(full)}</a>"
        )

        # Case manager display: Employee > User.full_name > user id
        cm_emp = (r.get("employee_full_name") or "").strip()
        cm_usr = (r.get("user_full_name") or "").strip()
        r["case_manager_name"] = cm_emp or cm_usr or (r.get("case_manager") or "")

        # Nicer snippet
        snippet = strip_html(r.get("summary_raw") or "")
        snippet = re.sub(r"\s+", " ", snippet).strip()
        if len(snippet) > 220:
            snippet = snippet[:220].rsplit(" ", 1)[0] + "…"
        r["summary_snippet"] = (
            f'<div class="text-muted small" style="border-left:3px solid #e2e6ea;padding-left:.5rem;">'
            f"{frappe.utils.escape_html(snippet)}</div>"
        )

    columns = [
        {"label": _("Date"), "fieldname": "entry_date_pretty", "fieldtype": "Data", "width": 150},
        {
            "label": _("Case"),
            "fieldname": "referral_case",
            "fieldtype": "Link",
            "options": "Referral Case",
            "width": 150,
        },
        {
            "label": _("Referral"),
            "fieldname": "referral",
            "fieldtype": "Link",
            "options": "Student Referral",
            "width": 150,
        },
        {"label": _("Student"), "fieldname": "student_link", "fieldtype": "HTML", "width": 220},
        {"label": _("School"), "fieldname": "school", "fieldtype": "Link", "options": "School", "width": 140},
        {"label": _("Program"), "fieldname": "program", "fieldtype": "Link", "options": "Program", "width": 140},
        {
            "label": _("Academic Year"),
            "fieldname": "academic_year",
            "fieldtype": "Link",
            "options": "Academic Year",
            "width": 140,
        },
        {"label": _("Case Manager"), "fieldname": "case_manager_name", "fieldtype": "Data", "width": 180},
        {"label": _("Case Severity"), "fieldname": "case_severity", "fieldtype": "Data", "width": 110},
        {"label": _("Case Status"), "fieldname": "case_status", "fieldtype": "Data", "width": 110},
        {"label": _("Entry Type"), "fieldname": "entry_type", "fieldtype": "Data", "width": 150},
        {"label": _("Assignee"), "fieldname": "assignee", "fieldtype": "Link", "options": "User", "width": 170},
        {"label": _("Entry Status"), "fieldname": "entry_status", "fieldtype": "Data", "width": 110},
        {"label": _("Summary"), "fieldname": "summary_snippet", "fieldtype": "HTML", "width": 420},
    ]

    # ── Charts (default = weekly) ───────────────────────────────────────
    chart_over_time_week = _chart_entries_over_time(rows, bucket="week")
    chart_over_time_month = _chart_entries_over_time(rows, bucket="month")
    chart_by_school = _chart_simple_count(rows, key="school", title=_("Entries by School"))
    chart_by_program = _chart_simple_count(rows, key="program", title=_("Entries by Program"))
    chart_by_manager = _chart_simple_count(rows, key="case_manager_name", title=_("Entries per Case Manager"))

    message = {
        "chart_over_time_week": chart_over_time_week,
        "chart_over_time_month": chart_over_time_month,
        "chart_by_school": chart_by_school,
        "chart_by_program": chart_by_program,
        "chart_by_manager": chart_by_manager,
    }

    # NEW: pick which chart to show based on the toggle
    selected_bucket = (filters.get("time_bucket") or "Week").strip().lower()
    selected_chart = chart_over_time_month if selected_bucket.startswith("m") else chart_over_time_week

    return columns, rows, message, selected_chart


# ---------------- helpers ----------------


def _guard_roles():
    user = frappe.session.user
    if user == "Administrator":
        return
    roles = set(frappe.get_roles(user))
    if not (roles & ALLOWED):
        frappe.throw(_("You are not permitted to run this report."))


def _build_where(f, date_from, date_to, site_tz):
    where, params = [], {}
    params["site_tz"] = site_tz
    if date_from:
        where.append(f"{ENTRY_LOCAL_EXPR} >= %(df)s")
        params["df"] = date_from  # noqa: E701, E702
    if date_to:
        where.append(f"{ENTRY_LOCAL_EXPR} <= %(dt)s")
        params["dt"] = date_to  # noqa: E701, E702
    if f.get("student"):
        where.append("rc.student = %(student)s")
        params["student"] = f.student  # noqa: E701, E702
    if f.get("referral"):
        where.append("rc.referral = %(referral)s")
        params["referral"] = f.referral  # noqa: E701, E702
    if f.get("school"):
        where.append("rc.school = %(school)s")
        params["school"] = f.school  # noqa: E701, E702
    if f.get("program"):
        where.append("rc.program = %(program)s")
        params["program"] = f.program  # noqa: E701, E702
    if f.get("academic_year"):
        where.append("rc.academic_year = %(ay)s")
        params["ay"] = f.academic_year  # noqa: E701, E702
    if f.get("case_manager"):
        where.append("rc.case_manager = %(cm)s")
        params["cm"] = f.case_manager  # noqa: E701, E702
    if f.get("assignee"):
        where.append("rce.assignee = %(assignee)s")
        params["assignee"] = f.assignee  # noqa: E701, E702
    if f.get("entry_type"):
        where.append("rce.entry_type = %(et)s")
        params["et"] = f.entry_type  # noqa: E701, E702
    if f.get("entry_status"):
        where.append("rce.status = %(es)s")
        params["es"] = f.entry_status  # noqa: E701, E702
    if f.get("case_severity"):
        where.append("rc.severity = %(sev)s")
        params["sev"] = f.case_severity  # noqa: E701, E702
    if f.get("case_status"):
        where.append("rc.case_status = %(cs)s")
        params["cs"] = f.case_status  # noqa: E701, E702
    return where, params


def _chart_entries_over_time(rows, bucket="week"):
    """Return a line chart; bucket = 'week'|'month'.
    Key the aggregation by real date objects → reliable sorting."""
    counts = defaultdict(int)

    def week_start(d: date) -> date:
        return d - timedelta(days=d.weekday())  # Monday

    for r in rows:
        d = r.get("entry_date")
        if not d:
            continue
        if bucket == "week":
            k = week_start(d)
        else:
            k = date(d.year, d.month, 1)
        counts[k] += 1

    if not counts:
        return None

    keys = sorted(counts.keys())
    if bucket == "week":
        labels = [formatdate(k, "d MMM yyyy") for k in keys]
    else:
        labels = [formatdate(k, "MMM yyyy") for k in keys]
    values = [counts[k] for k in keys]

    return {
        "data": {"labels": labels, "datasets": [{"name": _("Entries"), "values": values}]},
        "type": "line",
    }


def _chart_simple_count(rows, key, title):
    counts = defaultdict(int)
    for r in rows:
        val = (r.get(key) or "").strip() or _("(Missing)")
        counts[val] += 1
    if not counts:
        return None
    labels = sorted(counts.keys(), key=lambda s: s.lower())
    values = [counts[label] for label in labels]
    return {"data": {"labels": labels, "datasets": [{"name": title, "values": values}]}, "type": "bar"}
